evm_arithmetization/src/prover.rs

```diff
@@ -249,6 +340,1098 @@ type ProofWithMemCaps<F, C, H, const D: usize> = (
+    res.push(auxiliary_columns_single_stark::<F, C, CpuStark<F, D>, D>(
+        all_stark.cpu_stark,
+        config,
+        &trace_poly_values[*Table::Cpu],
+        &ctl_data_per_table[*Table::Cpu],
+        ctl_challenges,
+    ));
+
+    // Keccak.
+    res.push(
+        auxiliary_columns_single_stark::<F, C, KeccakStark<F, D>, D>(
+            all_stark.keccak_stark,
+            config,
+            &trace_poly_values[*Table::Keccak],
+            &ctl_data_per_table[*Table::Keccak],
+            ctl_challenges,
+        ),
+    );
+
+    // KeccakSponge.
+    res.push(auxiliary_columns_single_stark::<
+        F,
+        C,
+        KeccakSpongeStark<F, D>,
+        D,
+    >(
+        all_stark.keccak_sponge_stark,
+        config,
+        &trace_poly_values[*Table::KeccakSponge],
+        &ctl_data_per_table[*Table::KeccakSponge],
+        ctl_challenges,
+    ));
+
+    // Logic.
+    res.push(auxiliary_columns_single_stark::<F, C, LogicStark<F, D>, D>(
+        all_stark.logic_stark,
+        config,
+        &trace_poly_values[*Table::Logic],
+        &ctl_data_per_table[*Table::Logic],
+        ctl_challenges,
+    ));
+
+    // Memory.
+    res.push(
+        auxiliary_columns_single_stark::<F, C, MemoryStark<F, D>, D>(
+            all_stark.memory_stark,
+            config,
+            &trace_poly_values[*Table::Memory],
+            &ctl_data_per_table[*Table::Memory],
+            ctl_challenges,
+        ),
+    );
+
+    // MemBefore.
+    res.push(auxiliary_columns_single_stark::<
+        F,
+        C,
+        MemoryContinuationStark<F, D>,
+        D,
+    >(
+        all_stark.mem_before_stark,
+        config,
+        &trace_poly_values[*Table::MemBefore],
+        &ctl_data_per_table[*Table::MemBefore],
+        ctl_challenges,
+    ));
+
+    // MemAfter.
+    res.push(auxiliary_columns_single_stark::<
+        F,
+        C,
+        MemoryContinuationStark<F, D>,
+        D,
+    >(
+        all_stark.mem_before_stark,
+        config,
+        &trace_poly_values[*Table::MemAfter],
+        &ctl_data_per_table[*Table::MemAfter],
+        ctl_challenges,
+    ));
+
+    res
+}
+
+fn auxiliary_columns_single_stark<F, C, S, const D: usize>(
+    stark: S,
+    config: &StarkConfig,
+    trace_poly_values: &[PolynomialValues<F>],
+    ctl_data: &CtlData<F>,
+    ctl_challenges: &GrandProductChallengeSet<F>,
+) -> Vec<PolynomialValues<F>>
+where
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    S: Stark<F, D>,
+{
+    let rate_bits = config.fri_config.rate_bits;
+    let constraint_degree = stark.constraint_degree();
+    assert!(
+        constraint_degree <= (1 << rate_bits) + 1,
+        "The degree of the Stark constraints must be <= blowup_factor + 1"
+    );
+
+    let lookup_challenges: Vec<_> = ctl_challenges.challenges.iter().map(|ch| ch.beta).collect();
+    // Add lookup columns.
+    let lookups = stark.lookups();
+    let mut res = {
+        let mut columns = Vec::new();
```

Comment:
There are more columns than that, since the number of lookup columns depend on the individual lookups. I don't think computing it just for the capacity is worth it?

Reply:
yeah it's a lower bound, but it avoids all intermediary allocations. Right now we're starting with a capacity of 4.